disperser/meterer/meterer.go

```diff
@@ -0,0 +1,273 @@
+	// Create the EIP712Signer
+	//TODO: update the chainID and verifyingContract
+	signer := NewEIP712Signer(big.NewInt(17000), common.HexToAddress("0x1234000000000000000000000000000000000000"))
+
+	recoveredAddress, err := signer.RecoverSender(&header)
+	if err != nil {
+		return fmt.Errorf("failed to recover sender: %w", err)
+	}
+
+	accountAddress := common.HexToAddress(header.AccountID)
+
+	if recoveredAddress != accountAddress {
+		return fmt.Errorf("invalid signature: recovered address %s does not match account ID %s", recoveredAddress.Hex(), accountAddress.Hex())
+	}
+
+	return nil
+}
+
+// ServeReservationRequest handles the rate limiting logic for incoming requests
+func (m *Meterer) ServeReservationRequest(ctx context.Context, blobHeader BlobHeader, reservation *ActiveReservation) error {
+	if !m.ValidateBinIndex(blobHeader, reservation) {
+		return fmt.Errorf("invalid bin index for reservation")
+	}
+
+	// Update bin usage atomically and check against reservation's data rate as the bin limit
+	if err := m.IncrementBinUsage(ctx, blobHeader, reservation.DataRate); err != nil {
+		return fmt.Errorf("bin overflows: %w", err)
+	}
+
+	return nil
+}
+
+// ValidateBinIndex checks if the provided bin index is valid
+func (m *Meterer) ValidateBinIndex(blobHeader BlobHeader, reservation *ActiveReservation) bool {
+	currentBinIndex := GetCurrentBinIndex()
+	// Valid bin indexes are either the current bin or the previous bin
+	if (blobHeader.BinIndex != currentBinIndex && blobHeader.BinIndex != (currentBinIndex-1)) || (reservation.StartEpoch > blobHeader.BinIndex || blobHeader.BinIndex > reservation.EndEpoch) {
+		return false
+	}
+	return true
+}
+
+// IncrementBinUsage increments the bin usage atomically and checks for overflow
+// TODO: Bin limit should be direct write to the Store
+func (m *Meterer) IncrementBinUsage(ctx context.Context, blobHeader BlobHeader, binLimit uint32) error {
+	newUsage, err := m.OffchainStore.UpdateReservationBin(ctx, blobHeader.AccountID, uint64(blobHeader.BinIndex), blobHeader.BlobSize)
+	if err != nil {
+		return fmt.Errorf("failed to increment bin usage: %w", err)
+	}
+
+	// metered usage stays within the bin limit
+	if newUsage <= binLimit {
+		return nil
+	} else if newUsage-blobHeader.BlobSize >= binLimit {
+		// metered usage before updating the size already exceeded the limit
+		return fmt.Errorf("Bin has already been overflowed")
+	}
+
+	// check against the onchain state for overflow (TODO: update to real provider calls)
+	blockNumber := uint(0)
+	reservation, err := m.ChainState.MockedGetActiveReservationByAccount(ctx, blockNumber, blobHeader.AccountID)
+	if err != nil {
+		return fmt.Errorf("failed to get active reservation by account: %w", err)
+	}
+	if newUsage <= 2*binLimit && blobHeader.BinIndex+2 <= reservation.EndEpoch {
+		m.OffchainStore.UpdateReservationBin(ctx, blobHeader.AccountID, uint64(blobHeader.BinIndex+2), newUsage-binLimit)
```

Comment:
Note: this is not a race condition because previous logic ensures this can only happen in a single thread.